### ISSUE TITLE
Add Linux support via `dasht`

### ISSUE THREAD

**Issue by mrjones2014** [mrjones2014/dash.nvim#24]
https://github.com/sunaku/dasht

A Linux client to download and search docsets, just like Dash, but without needing the actual Dash.app since that's a macOS app.

We can probably detect whether we're on macOS or Linux, and use a different backend on Linux. Likely will need a separate Rust backend for Linux.

This will also require tweaking the config a bit I think.

**Comment by mrjones2014** [comment]
Ohhh, okay, gotcha @oblitum 

>relies on Dash docsets

that's what I liked about `dasht`. If Zeal is the same way, I can work on adding support for Zeal to this plugin then. I've updated the issue description.

**Comment by mrjones2014** [comment]
@oblitum does Zeal have a CLI? For the Dash integration I'm using the same CLI that powers their Alfred integration.

**Comment by oblitum** [comment]
@mrjones2014 I never used, my sole usage or it, for years, is just by launching the gui, either directly or from [n]vim with zeavim.

**Comment by smithbm2316** [comment]
This is most definitely a separate plugin idea, but it would be an interesting to me to just build a plugin that happens to *use* the Dash docsets, and renders them within a new Neovim Filetype, similarly to how you can directly navigate the Vim help docs natively. I can't speak for everyone, but while I really like Dash.app, I mostly like it because it's one single repository essentially of documentation that is easily searchable. I'd much rather have access to those docs within Neovim itself rather than have to go to a separate app personally, because then I can scroll, search, query, and navigate documentation for anything I want using my Neovim keybindings. 

Might be an interesting idea to explore just setting up almost a "Neovim Client" for Dash docsets where you generate the SQLite file with the plugin and add/remove docsets from within the plugin itself. Now that's a lot of just re-doing what Dash.app or Zeal or dasht already do, but it could be a fun plugin idea and give us as the Neovim users much more control of the end user experience. Just kinda spitballing here, it might be something I'd be interested in hacking together myself. Thanks for all of the awesome work you've done on this plugin so far, it's really excellent!

**Comment by lukelex** [comment]
Does anyone know where the SQLite file lives? I'd love to take a peak at how things are structure there.

**Comment by smithbm2316** [comment]
> Does anyone know where the SQLite file lives? I'd love to take a peak at how things are structure there.

There's a SQLite file afaik for *each* docset you install. It's not one big file with all the docsets in it

**Comment by mrjones2014** [comment]
@smithbm2316 I’d be more than open to implementing that here, but the issue I’m running into is that the contents are HTML files.

I don’t know of a good way to render the HTML within Neovim. I’ve tried w3m and browsh (text-based terminal browsers) but neither looks very good.

See #3

**Comment by smithbm2316** [comment]
> @smithbm2316 I’d be more than open to implementing that here, but the issue I’m running into is that the contents are HTML files. I don’t know of a good way to render the HTML within Neovim.

I wonder if there would be an easy way to integrate an `html -> markdown` converter/compiler and then just render the markdown inside of Neovim. Something like [this](https://github.com/JohannesKaufmann/html-to-markdown) as an optional build step when `dash.nvim` is installed, but perhaps with a Rust-based package that does that job. Again, very much just _spitballing_ here. This is most likely some serious overengineering lol. Food for thought though. I'll have a think about this HTML rendering issue in the coming week, I forgot that the underlying docsets are HTML and not something easily human-readable.

**Comment by xbc5** [comment]
I've been thinking about this idea for a few months now, and found it because I was looking for exactly this. Being able to just fuzzy find over docs without leaving Neovim is a fantastic idea. I agree with smithb2316, en entirely pure Neovim experience would be great.

I will watch this closely, and contribute where I feel that I can.

**Comment by mrjones2014** [comment]
@xbc5 I’m concerned about having to maintain markdown versions of docsets myself. But this idea is more related to #3 anyway, it doesn’t help with fuzzy finding them.

**Comment by xbc5** [comment]
@mrjones2014

1. couldn't I fuzzy-find over the produced markdown? It might not be plain old text, but is it good enough?
2. you wouldn't have to maintain markdown versions of the docsets, just the translator (i.e. ensuring *clean* translation) -- users can translate when they install a docset (it's fast);

Are you looking for plain old text? There's code/syntax highlighting to consider, which is most important in this context -- everything else is secondary.

**Comment by mrjones2014** [comment]
Yeah, but translating from HTML to Markdown is much, much harder than translating from Markdown to HTML. The input space is far greater. Converting HTML to Markdown is far beyond the scope of this project.

**Comment by mosaleh52** [comment]
@mrjones2014 I have read the discussion and what I have understand that the problem now is rendering HTML inside neovim and the db problems have been resolved ,am I right

**Comment by teto** [comment]
there is https://github.com/tami5/sqlite.lua too that could be a native lua implementation

**Comment by mrjones2014** [comment]
Yeah, we'd want to do all our db interactions from the Rust backend though, otherwise we'd be making two round trips from lua -> Rust -> lua -> Rust -> lua for every db interaction

**Comment by mosaleh52** [comment]
@mrjones2014 I have basic knowledge writing sql queries , is there any way that i could help 
with some direction i will try to find sql related solution

**Comment by mrjones2014** [comment]
@mosaleh52  You can try figuring out what the data schema looks like for the sqlite databases.

**Comment by mrjones2014** [comment]
If it’s the same as the Dash docset dbs, then yeah. I think they’re the same but I’m not sure.

**Comment by vihu** [comment]
Have been following this issue for a while now and was wondering if we could perhaps split it into potentially doable tasks?

Just for information, there have been a few recent interesting developments for neovim support with rust, specifically [nvim-oxy](https://github.com/noib3/nvim-oxi) and combine that with [sqlite.lua](https://github.com/kkharji/sqlite.lua) bindings maybe we could get the ball rolling?

**Comment by mrjones2014** [comment]
Yeah I recently discovered `nvim-oxi` — using it would simplify the existing Rust code base quite a bit, as well as potentially make this task quite a bit easier.

If introducing `nvim-oxi` I’d be inclined to move _all_ the code to Rust, and use a Rust crate for SQLite instead.

A first step would be adding nvim-oxi, removing `mlua`, which will probably require a lot of Rust refactoring and rewriting all the Lua code into Rust.

**Comment by mrjones2014** [comment]
Although additionally, with enhancements to nvim itself, there’s a lot of other things we could do to make the whole plugin a lot simpler. It may be worth doing a ground up rewrite using `nvim-oxi` and [rusqlite](https://github.com/rusqlite/rusqlite)

**Comment by vihu** [comment]
> Although additionally, with enhancements to nvim itself, there’s a lot of other things we could do to make the whole plugin a lot simpler. It may be worth doing a ground up rewrite using `nvim-oxi` and [rusqlite](https://github.com/rusqlite/rusqlite)

A rewrite from the ground up purely in rust might definitely be worth it. Just one suggestion (from personal XP), instead of rusqlite, [sqlx](https://github.com/launchbadge/sqlx) with `sqlite` feature flag might also serve as an alternative.

Another nice thing about doing it all (from the ground up) in rust is that we could potentially just do the docset binding stuff first and leave nvim bindings entirely (for first pass), nvim-oxi can follow later once we get to that point.

**Comment by mrjones2014** [comment]
Agreed. I don't have a ton of time to work on this plugin anymore so any PRs are super appreciated.
